### PR TITLE
Avro support in Kafka connector

### PR DESCRIPTION
**Pull request by anusudarsan**

### CODE REVIEW

File: presto-record-decoder/src/test/java/com/facebook/presto/decoder/avro/TestAvroDecoder.java [hunk 1]
```diff
@@ -79,23 +81,42 @@
 
     private static String getAvroSchema(String name, String dataType)
     {
+        return getAvroSchema(ImmutableMap.of(name, dataType));
+    }
+
+    private static String getAvroSchema(Map<String, String> fields)
+    {
+        // the default value is set to null
```
Comment (findepi) [inline presto-record-decoder/src/test/java/com/facebook/presto/decoder/avro/TestAvroDecoder.java]: unnecessary

File: presto-product-tests/conf/presto/etc/config.properties [hunk 2]
```diff
@@ -37,6 +37,7 @@ plugin.bundles=\
   ../../../presto-postgresql/pom.xml,\
   ../../../presto-sqlserver/pom.xml,\
   ../../../presto-ml/pom.xml,\
+  ../../../presto-kafka/pom.xml,\
```
Comment (findepi) [inline presto-product-tests/conf/presto/etc/config.properties]: revert?

**Comment by anusudarsan** [comment]
@findepi ac

File: presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaSplitManager.java [hunk 3]
```diff
@@ -129,6 +138,22 @@ public ConnectorSplitSource getSplits(ConnectorTransactionHandle transaction, Co
         }
     }
 
+    private static String readSchema(String dataSchemaLocation)
+    {
+        try {
+            InputStream inputStream = isURL(dataSchemaLocation.trim().toLowerCase(Locale.ENGLISH)) ? new URL(dataSchemaLocation).openStream() : new FileInputStream(new File(dataSchemaLocation));
```
Comment (nezihyigitbasi) [inline presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaSplitManager.java]: Make sure to close `inputStream`, I don't think it's closed now.

Comment (nezihyigitbasi) [inline presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaSplitManager.java]: - static import `ENGLISH`
- `new FileInputStream(dataSchemaLocation)` is more concise
- add `e` as the cause to the `PrestoException` on L148

File: presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaSplitManager.java [hunk 4]
```diff
@@ -129,6 +138,22 @@ public ConnectorSplitSource getSplits(ConnectorTransactionHandle transaction, Co
...
+            return CharStreams.toString(new InputStreamReader(inputStream, UTF_8));
+        }
+        catch (IOException e) {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, "Could not parse the AVRO schema at: " + dataSchemaLocation);
+        }
+    }
+
+    private static boolean isURL(String location)
```
Comment (nezihyigitbasi) [inline presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaSplitManager.java]: Is there a reason why you don't use the `URI` constructor and handle the exception if it's not a url, etc.?

File: presto-product-tests/src/main/java/com/facebook/presto/tests/kafka/KafkaAvroSmokeTest.java [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+        }
+    }
+
+    private static byte[] convertRecordToAvro(Schema schema, Map<String, Object> values)
+    {
+        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+        GenericData.Record record = new GenericData.Record(schema);
+        values.forEach(record::put);
+        try {
+            DataFileWriter<GenericRecord> dataFileWriter = new DataFileWriter<>(new GenericDatumWriter<>(schema));
```
Comment (nezihyigitbasi) [inline presto-product-tests/src/main/java/com/facebook/presto/tests/kafka/KafkaAvroSmokeTest.java]:
```
try (DataFileWriter<GenericRecord> dataFileWriter = new DataFileWriter<>(new GenericDatumWriter<>(schema))) {
```

File: presto-product-tests/src/main/java/com/facebook/presto/tests/kafka/KafkaAvroSmokeTest.java [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+                    new ListKafkaDataSource(ImmutableList.of(
+                            new KafkaMessage(
+                                    contentsBuilder()
+                                            .appendBytes(avroData)
+                                            .build()))),
+                    1,
+                    1));
+        }
+        catch (IOException e) {
+            throw new RuntimeException(e);
```
Comment (nezihyigitbasi) [inline presto-product-tests/src/main/java/com/facebook/presto/tests/kafka/KafkaAvroSmokeTest.java]: You can use `UncheckedIOException`.

File: presto-record-decoder/pom.xml [hunk 7]
```diff
@@ -51,6 +51,19 @@
             <artifactId>joda-time</artifactId>
         </dependency>
 
+        <dependency>
+            <groupId>org.apache.avro</groupId>
```
Comment (nezihyigitbasi) [inline presto-record-decoder/pom.xml]: We put versions in the top level pom.

File: presto-record-decoder/src/main/java/com/facebook/presto/decoder/avro/AvroRowDecoder.java [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+    @Override
+    public Optional<Map<DecoderColumnHandle, FieldValueProvider>> decodeRow(byte[] data, Map<String, String> dataMap)
+    {
+        GenericRecord avroRecord;
+        try {
+            // Assumes producer uses DataFileWriter or data comes in this particular format.
+            // TODO: Support other forms for producers
+            DataFileStream<GenericRecord> dataFileReader = new DataFileStream<>(new ByteArrayInputStream(data), avroRecordReader);
```
Comment (nezihyigitbasi) [inline presto-record-decoder/src/main/java/com/facebook/presto/decoder/avro/AvroRowDecoder.java]: `DataFileStream` is closeable, I guess you should close it.

File: presto-redis/src/main/java/com/facebook/presto/redis/RedisDecoderModule.java [hunk 9]
```diff
@@ -49,6 +51,7 @@ public void configure(Binder binder)
         decoderFactoriesByName.addBinding(RawRowDecoder.NAME).to(RawRowDecoderFactory.class).in(SINGLETON);
         decoderFactoriesByName.addBinding(ZsetRedisRowDecoder.NAME).to(ZsetRedisRowDecoderFactory.class).in(SINGLETON);
         decoderFactoriesByName.addBinding(HashRedisRowDecoder.NAME).to(HashRedisRowDecoderFactory.class).in(SINGLETON);
+        decoderFactoriesByName.addBinding(AvroRowDecoder.NAME).to(AvroRowDecoderFactory.class).in(SINGLETON);
```
Comment (nezihyigitbasi) [inline presto-redis/src/main/java/com/facebook/presto/redis/RedisDecoderModule.java]: It would be better if you make this change in another commit as this is changing `presto-redis`, but the commit message says `Add basic Avro decoder for Kafka`.

File: presto-main/src/main/java/com/facebook/presto/testing/TestingEnv.java [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+import com.facebook.presto.sql.analyzer.FeaturesConfig;
+import com.facebook.presto.type.TypeRegistry;
+
+public class TestingEnv
+{
+    private TestingEnv() {}
+
+    public static final TypeManager TYPE_MANAGER = new TypeRegistry();
+    static {
+        // wire TYPE_MANAGER with function retistry
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/testing/TestingEnv.java]: Typo "retistry"

File: presto-main/src/main/java/com/facebook/presto/testing/TestingEnv.java [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+ */
+package com.facebook.presto.testing;
+
+import com.facebook.presto.block.BlockEncodingManager;
+import com.facebook.presto.metadata.FunctionRegistry;
+import com.facebook.presto.spi.type.TypeManager;
+import com.facebook.presto.sql.analyzer.FeaturesConfig;
+import com.facebook.presto.type.TypeRegistry;
+
+public class TestingEnv
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/testing/TestingEnv.java]: Let's call this `TestingEnvironment`.

File: presto-record-decoder/pom.xml (hunk 7 above)
Comment (electrum) [reply presto-record-decoder/pom.xml]: It might be appropriate to do here if we really need a different version of Avro between the Hive and Kafka connectors, though it's preferred to use the `dependencyManagement` version in the root POM.

**Comment by findepi** [comment]
@electrum @nezihyigitbasi any further comments? Or we should just merge this?

**Comment by cricket007** [comment]
Any plans on build a Connector that uses a Schema Registry encoded Kafka payload rather than requiring schema to be emdedded? 

e.g Confluent's or Hortonwork's?

**Comment by dain** [comment]
I've never heard of that, but a good start would be to file a new issue with a link to the relevant documentation for the registry.  As for plans, I haven't heard anyone talking about that, so, it might be up to you to implement :)

**Comment by cricket007** [comment]
Done! #11354
